Removed memcpy that cause inconsistency in nbr-table when adding nd6 neighbors.

joakimeriksson opened the pull request:
This is a fix for ND6 that makes the nbr-table have two entries with the same link layer address. This cause permant disconnect from RPL root (until global repair or reboot). The reason it happens is that the nbr table allow adding a NULL entry for the neighbor discovery to handle the case when there is not yet any know link layer address. The NA code then just do a memcopy so that NULL suddenly is a valid link layer address and when that happen to exist in the tables already - things break down. This fix removes the NULL entry and re-installs it again to avoid the problem.

This bug was found by the Intel IoT team in Malaysia that have been testing the 6LoWPAN stack significantly so many thanks for your efforts! (@jeremy-ang, @usmansarwar and @pedquarkbld).

simonduq commented:
Great, I think this is the same issue as https://github.com/contiki-os/contiki/issues/1277
Two comments:
* I believe you can now remove the unsafe cast `uip_ds6_nbr_get_ll` to non const, i.e. declare `lladdr` as a const *, as discussed in https://github.com/contiki-os/contiki/issues/1277
* Please use `extract_lladdr_aligned` instead of casting `nd6_opt_llao` to `(const uip_lladdr_t *)`, see https://github.com/contiki-os/contiki/pull/1207

joakimeriksson commented:
It is exactly a fix for that issue and specifically the NULL case which is really happening in real-world cases (which maybe changing MAC address do typically not). I will try to fix the PR according to your comments and update it ASAP.

joakimeriksson commented:
Ok, that was a bit more changes than I expected - I changed to aligned now - which also ended up forcing me to change the other memcpy that might cause inconsistency (not that it ever happen in a normal case but - now it is also fixed in the same way as the NULL case).

simonduq commented:
:+1:

nfi commented on core/net/ipv6/uip-nd6.c:
```diff
@@ -489,23 +489,32 @@ na_input(void)
       is_llchange =
-        memcmp(&nd6_opt_llao[UIP_ND6_OPT_DATA_OFFSET], (void *)lladdr,
+        memcmp(&nd6_opt_llao[UIP_ND6_OPT_DATA_OFFSET], (void *)&lladdr_aligned,
                UIP_LLADDR_LEN);
```
With the latest update, lladdr_aligned will always set to &nd6_opt_llao[UIP_ND6_OPT_DATA_OFFSET] and not NBR lladdr so the compare will always say no change.

simonduq replied:
oops I missed that, needs to compare against `uip_ds6_nbr_get_ll`

joakimeriksson commented:
Good catch... I will fix now!

joakimeriksson commented:
It is now fixed and ready to go!

simonduq commented on core/net/ipv6/uip-nd6.c:
```diff
@@ -489,23 +489,32 @@ na_input(void)
+    if(nd6_opt_llao != NULL) {
       is_llchange =
-        memcmp(&nd6_opt_llao[UIP_ND6_OPT_DATA_OFFSET], (void *)lladdr,
+        memcmp(&nd6_opt_llao[UIP_ND6_OPT_DATA_OFFSET], uip_ds6_nbr_get_ll(nbr),
```
I think you actually need to first get the lladdr, then check it is non-null before `memcmp`

joakimeriksson replied:
I can add that so that if we have an invalid NBR it will even work (if the nbr is from the nor-module it will never have a null entry - yet).

simonduq replied:
yes! (better than memcmp on NULL, which could fail on MCUs with memory protection)

joakimeriksson replied:
Yes - which will soon happen when we remove the NULL MAC address thing in the nor-module ;-)

simonduq commented on core/net/ipv6/uip-nd6.c:
```diff
@@ -489,23 +489,34 @@ na_input(void)
-    if(nd6_opt_llao != 0) {
+
+    lladdr = uip_ds6_nbr_get_ll(nbr);
+    if(nd6_opt_llao != NULL && lladdr != NULL) {
```
don't we want to set `is_llchange` if `lladdr` is `NULL`?

joakimeriksson commented:
Ok - so to sum it up. I added the null check on lladdr base on @simonduq's comment and also fixed so that we now flush the NBR cache whenever the override flag is set in the NA packet. Also we do not flush the NBR cache when the override flag is not set and there is no change to the lladdr. Before there was a risk of updating the lladdr with a memcpy with the same lladdr - and with my fix it flushed the cache instead of doing nothing when getting a non changed lladdr and a non-override. Note: before we just updated lladdr and never flushed the NBR cache (which the RFC says should be done on override).

joakimeriksson commented:
I finally moved the change of lladdr into the nbr-module in order to get a consistent model of how to update the lladdr for a neighbor cache entry when the lladdr is changed (either from null to a real value at INCOMPLETE state or from one to another MAC address in other states).

joakimeriksson commented:
Ping!  Are we ok with having this bug still around? Or is it now solved by other fixes?

simonduq commented on core/net/nbr-table.c:
```diff
@@ -169,6 +169,26 @@ nbr_set_bit(uint8_t *bitmap, nbr_table_t *table, nbr_table_item_t *item, int val
+  used_map[index_from_key(least_used_key)] = 0;
+  /* Remove neighbor from list */
+  list_remove(nbr_table_keys, least_used_key);
+  /* Return associated key */
```
the above comment should be removed

simonduq commented on core/net/nbr-table.c:
```diff
@@ -169,6 +169,26 @@ nbr_set_bit(uint8_t *bitmap, nbr_table_t *table, nbr_table_item_t *item, int val
+  /* Return associated key */
+}
+
 static nbr_table_key_t *
```
for consistency I would add `/*---------------------------------------------------------------------------*/` here

simonduq commented:
looks good but I'll need to take a closer look, will do asap

simonduq commented:
Ready to be merged after my cosmetic comments are addressed (and after some timeout)

joakimeriksson commented:
Did a rebase - and I guess the timeout should be triggered already!

simonduq commented:
ah, forgot the timeout tag! let me add

simonduq commented:
as there was no timeout label, let's give this PR another couple days before merging

joakimeriksson commented:
Ok - I will be picky next time the tag is not there ;-)

nvt commented:
:+1: after @joakimeriksson makes some minor updates discusses offline.

simonduq commented:
are the updates done?

joakimeriksson commented:
I think that was exactly what I did last commit - so yes!

simonduq commented:
merging after timeout!
